### PR TITLE
Py3: common code migration

### PR DESCRIPTION
**Pull request by klaussfreire**
Make an initial push toward Python 3 support:

- Fix setup.py to run in py3
- Remove dual python support idioms
- Fix syntax errors due to removed syntax from py2 (except and a few other 2to3 transforms)
- Import renames
- Switch cPickle, cStringIO and friends to py3 equivalents without py2 fallback
- Basic unicode/bytes separation (very basic)
- keys/iterkeys and friends, but only for dicts, not for caches (maybe do caches in the future?)
- Division semantics fixed in some places (not thorough though)

This makes tests at least run.

Tests that seem to pass:

- inproc
- tiered
- external_integration
- threadpooltest (almost, 2 failures that don't seem related to py3)

Everything else is failing with various errors that will be addressed on a followup PR.

The decision to drop py2 compatibility simplified many things that I remember were a lot more painful in the previous attempt.

### CODE REVIEW

File: lib/chorde/_version.py [hunk 1]
```diff
@@ -1 +1 @@
-__version__ = '0.8.4'
\ No newline at end of file
+__version__ = '0.9.0'
```
Comment (gefarion) [inline lib/chorde/_version.py:1]: Why not release the `1.0.0` version?

File: lib/chorde/clients/memcached.py [hunk 2]
```diff
@@ -1031,7 +1024,7 @@ def shorten_key(self, key,
                 zpfx = self.compress_prefix
             except:
                 # Try pickling
-                key = "P#"+self.key_pickler.dumps(key,2).encode("base64").replace("\n","")
+                key = "P#"+b64encode(self.key_pickler.dumps(key,2))
```
Comment (gefarion) [inline lib/chorde/clients/memcached.py:1027]: Don't you need to remove the `\n`?

File: tests/files.py [hunk 3]
```diff
@@ -133,7 +133,7 @@ def testLimit(self):
         bigval = "12" * chorde.clients.files.MMAP_THRESHOLD
         maxentries = SIZE / len(bigval)
```
Comment (gefarion) [inline tests/files.py:134]: Don't you have to use `//` here?

File: tests/threadpooltest.py [hunk 4]
```diff
@@ -131,15 +131,15 @@ def testConcurrency(self):
...
                 self.pool.apply_async(accounting, (i,))
-        threads = [ Thread(target=killit, args=(i,)) for i in xrange(M) ]
+        threads = [ Thread(target=killit, args=(i,)) for i in range(M) ]
         for t in threads:
             t.start()
         for t in threads:
             t.join()
         self.join_continue(self.pool, 60)
-        total_counts = self.pool.apply(sum, (counts.itervalues(),))
+        total_counts = self.pool.apply(sum, (iter(counts.values()),))
```
Comment (gefarion) [inline tests/threadpooltest.py:142]: Don't `values` already return an iterator?

File: lib/chorde/_version.py (hunk 1 above)
Comment (klaussfreire) [reply lib/chorde/_version.py:1]: This is a "migrate to py3" version, with not many functional changes. I feel a 1.0.0 release should have some significant functional changes to merit the "promotion".

File: lib/chorde/clients/memcached.py (hunk 2 above)
Comment (klaussfreire) [reply lib/chorde/clients/memcached.py:1027]: b64encode doesn't add them (`str.encode("base64")` did)

File: tests/threadpooltest.py (hunk 4 above)
Comment (klaussfreire) [reply tests/threadpooltest.py:142]: Only if they're dicts. This was done by 2to3 and I bet it didn't know defaultdict acts like a dict in this case.

File: lib/chorde/_version.py (hunk 1 above)
Comment (gefarion) [reply lib/chorde/_version.py:1]: Deprecate py2 i think that merit the promotion.

**Review by gefarion — APPROVED** [review]

File: lib/chorde/decorators.py [hunk 5]
```diff
@@ -78,12 +79,12 @@ def _make_namespace(f, salt = None, salt2 = None):
             body_digest.update(salt2)
         if fcode:
             body_digest.update(getattr(fcode, 'co_code', ''))
-        return "%s.%s#%s" % (mname,fname,body_digest.digest().encode("base64").strip("=\n"))
+        return "%s.%s#%s" % (mname,fname,b64encode(body_digest.digest()).strip("=\n"))
```
Comment (chpedersen) [inline lib/chorde/decorators.py:82]: Is the `strip` necessary here?

File: lib/chorde/clients/asyncache.py [hunk 6]
```diff
@@ -259,7 +259,7 @@ def _writer(self, key, reentrant = True):
                 kev = None
             del w
 
-            if thread_id is not None and thread_id not in map(operator.itemgetter(0), self.workset.values()):
+            if thread_id is not None and thread_id not in list(map(operator.itemgetter(0), list(self.workset.values()))):
```
Comment (chpedersen) [inline lib/chorde/clients/asyncache.py:262]: I think the `list`s are not necessary here, are they?

File: lib/chorde/clients/memcached.py [hunk 7]
```diff
@@ -535,13 +529,13 @@ def set_multi(self, mapping, time=0, key_prefix='', min_compress_len=0):
 
             self._statlog('set_multi')
 
-            server_keys, prefixed_to_orig_key = self._map_and_prefix_keys(mapping.iterkeys(), key_prefix)
+            server_keys, prefixed_to_orig_key = self._map_and_prefix_keys(iter(mapping.keys()), key_prefix)
```
Comment (chpedersen) [inline lib/chorde/clients/memcached.py:532]: extra `iter`

**Review by chpedersen — APPROVED** [review]
Looks ok. Maybe some the `list(xxx)` are not really necessary but they're way too many to keep track

File: lib/chorde/decorators.py (hunk 5 above)
Comment (klaussfreire) [reply lib/chorde/decorators.py:82]: Yes, if we want to generate the same namespaces. Could be an rstrip though.

**Pull request merged**